src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -0,0 +1,126 @@
+{
+    /// <summary>
+    /// Resolves references for a package by looking in local nuget and artifact
+    /// directories for Roslyn
+    /// </summary>
+    internal class LocalReferenceResolver : IMetadataReferenceResolver
+    {
+        private readonly Dictionary<Guid, string> _cache = new Dictionary<Guid, string>();
+        private readonly HashSet<DirectoryInfo> _indexDirectories = new HashSet<DirectoryInfo>();
+        private readonly ILogger _logger;
+
+        public LocalReferenceResolver(ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger<LocalReferenceResolver>();
+            _indexDirectories.Add(GetArtifactsDirectory());
+            _indexDirectories.Add(GetNugetCacheDirectory());
+        }
+
+        public static DirectoryInfo GetNugetCacheDirectory()
+        {
+            var nugetPackageDirectory = Environment.GetEnvironmentVariable("NUGET_PACKAGES");
+            if (nugetPackageDirectory is null)
+            {
+                nugetPackageDirectory = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".nuget");
+            }
+
+            return new DirectoryInfo(nugetPackageDirectory);
+        }
+
+        public Task<MetadataReference> ResolveReferenceAsync(MetadataReferenceInfo referenceInfo)
+        {
+            var path = Search(referenceInfo.Mvid, referenceInfo.Name);
+            return Task.FromResult<MetadataReference>(MetadataReference.CreateFromFile(path));
+        }
+
+        public Task<ImmutableArray<MetadataReference>> ResolveReferencesAsync(IEnumerable<MetadataReferenceInfo> references)
+        {
+            var referenceArray = references.ToImmutableArray();
+            CacheNames(referenceArray);
+
+            var files = referenceArray.Select(r => _cache[r.Mvid]);
+
+            var metadataReferences = files.Select(f => MetadataReference.CreateFromFile(f)).Cast<MetadataReference>().ToImmutableArray();
+            return Task.FromResult(metadataReferences);
+        }
+
+        public string Search(Guid mvid, string name)
+        {
+            if (_cache.TryGetValue(mvid, out var referencePath))
+            {
+                return referencePath;
+            }
+
+            foreach (var directory in _indexDirectories)
+            {
+                var foundFile = directory.GetFiles(name, SearchOption.AllDirectories).FirstOrDefault();
+                if (foundFile is null)
+                {
+                    continue;
+                }
+
+                referencePath = foundFile.FullName;
+                _cache[mvid] = referencePath;
+
+                _logger.LogTrace($"Caching [{mvid}, {referencePath}]");
+                return referencePath;
+            }
+
+            throw new KeyNotFoundException($"[{mvid}, {name}]");
+        }
+
+        public void CacheNames(ImmutableArray<MetadataReferenceInfo> names)
+        {
+            var remaining = new HashSet<MetadataReferenceInfo>(names);
```

Comment:
The type `MetadataReferenceInfo` is not valid as a key because it does not have correct equality semantics. Consider for example that it has an `ImmutableArray<string>` as a field for extern aliases and that type does not implement `IEquatable<T>`. Hence two references which have the same set of extern aliases will never compare equals here.

Reply:
Good catch. I don't think this optimization is worth being too complex with the code, I'll just keep the original array to search.